### PR TITLE
Slurm scheduler

### LINKED ISSUE #12: Use testcontainers
Currently you need to have a PostgreSQL instance running before running tests. Using https://testcontainers-python.readthedocs.io/en/latest/ we could provision a PostgreSQL Docker container in a pytest fixture.

Pro:
- Can no longer forget to start a PostgreSQL instance
- No chance of polluting the wrong database like production

Con:
- Requires Docker
- Need to download PostgreSQL Docker image

### PR DESCRIPTION
**Pull request by sverhoeven**
Adds Slurm scheduler support. 

Also adds
* sftp file system to copy input/output to/from a remote Slurm server.
* Testing against slurm is done using a Docker container. using https://github.com/testcontainers/testcontainers-python (fixes #12)
* Disable more flake8 rules which are false positives or not valid in tests.

To test spin up a Slurm container with
```shell
docker run --detach --publish 10022:22 xenonmiddleware/slurm:20
```

To submit a job in a Python REPL (`python -m asyncio` or `ipython`)

```python
from pathlib import Path
from bartender.filesystems.sftp import SftpFileSystem
from bartender.schedulers.abstract import JobDescription
from bartender.schedulers.slurm import SlurmScheduler
from bartender.schedulers.runner import SshCommandRunner

job_dir = Path('/tmp/jobs/myjob')
job_dir.mkdir(parents=True, exist_ok=True)
config = {
            "hostname": 'localhost',
            "port": 10022,
            "username": 'xenon',
            "password": 'javagat',
        }
scheduler = SlurmScheduler(runner=SshCommandRunner(config))
description = JobDescription(command="echo -n hello", job_dir=str(job_dir))
fs = SftpFileSystem(entry='/home/xenon', config=config)
localized_description = fs.localize_description(description, job_dir.parent)

await fs.upload(description, localized_description)

job_id = await scheduler.submit(localized_description)

print(await scheduler.state(job_id))
# Repeat fetching state until it is 'ok' or 'error'

await fs.download(localized_description, description)
```
After download should have stderr.txt, stdout.txt, returncode files in /tmp/jobs/myjob.

As similar job submission is done at https://github.com/i-VRESSE/bartender/blob/slurm-scheduler/bartender/tests/schedulers/test_slurm.py#L60

Don't forget to `docker rm -f <slurm container name or id>` when you are done.

### CODE REVIEW

**Comment by Peter9192** [comment]
Just a heads up that I successfully managed to execute the instructions in the top post. I'm still trying to grasp more fully how this works, but that also requires looking at #15.

File: bartender/_ssh_utils.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    """Connect to a host using SSH.
+
+    :param config: Configuration.
+    :return: The connection.
+    """
+    conn_vargs = {
+        "known_hosts": None,
+    }
+    if config["password"] is not None:
+        conn_vargs["agent_path"] = None
```
Comment (Peter9192) [inline bartender/_ssh_utils.py]: I guess setting `known_hosts` to None ignores the confirmation prompt for new connections. What does the `agent_path` setting do?

File: bartender/_ssh_utils.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
+async def ssh_connect(config: SshConnectConfig) -> SSHClientConnection:
+    """Connect to a host using SSH.
+
+    :param config: Configuration.
+    :return: The connection.
+    """
+    conn_vargs = {
+        "known_hosts": None,
```
Comment (sverhoeven) [inline bartender/_ssh_utils.py:22]:
```suggestion
        # disable server host key validation
        "known_hosts": None,
```

File: bartender/filesystems/local.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        :param description: A job description.
+        :param entry: The path to replace with the entry path of this file system.
+            For example given a file system with entry path of /remote/jobs
+            and given job_dir in job description of /local/jobs/myjobid
+            and given entry of /local/jobs
+            will return description with job dir /remote/jobs/myjobid .
+        :return: A job description local to this file system.
+        """
+        # TODO do something with entry argument?
+        return description
```
Comment (Peter9192) [inline bartender/filesystems/local.py:26]: Is this also intended to "translate" the path from a remote jobdescription to a corresponding jobdescription on the local fs?

File: bartender/schedulers/slurm.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            args=[],
+            cwd=description.job_dir,
+            stdin=script,
+        )
+        if returncode != 0:
+            raise RuntimeError(
+                f"Error running sbatch, exited with {returncode}: {stderr}",
+            )
+
+        return stdout.strip().split(" ")[-1]
```
Comment (Peter9192) [inline bartender/schedulers/slurm.py:75]:
```suggestion

        #  Get the "4" out of string like "Submitted batch job 4"
        return stdout.strip().split(" ")[-1]
```

File: bartender/tests/schedulers/test_slurm.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+        jid = await scheduler.submit(localized_description)
+
+        await wait_for_job(scheduler, jid)
+
+        await fs.download(localized_description, description)
+
+        assert (job_dir / "returncode").read_text() == "0"
+        assert (job_dir / "stdout.txt").read_text() == "hello"
+        assert (job_dir / "stderr.txt").read_text() == ""
```
Comment (Peter9192) [inline bartender/tests/schedulers/test_slurm.py:138]: Assert that there are no other files?

**Review by Peter9192 — APPROVED** [review]
Tested and pair-reviewed!

File: bartender/_ssh_utils.py (hunk 1 above)
Comment (sverhoeven) [reply bartender/_ssh_utils.py]: At https://github.com/i-VRESSE/bartender/blob/3bd16de1a2ed0a338067aeee97faf6ea8a1a75e2/src/bartender/ssh_utils.py#L32 has been corrected to not use ssh-agent when a password is supplied.

File: bartender/tests/schedulers/test_slurm.py (hunk 5 above)
Comment (sverhoeven) [reply bartender/tests/schedulers/test_slurm.py:138]: done

File: bartender/filesystems/local.py (hunk 3 above)
Comment (sverhoeven) [reply bartender/filesystems/local.py:26]: Postponed to https://github.com/i-VRESSE/bartender/issues/34

File: bartender/filesystems/abstract.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+from pathlib import Path
+from typing import Protocol
+
+from bartender.schedulers.abstract import JobDescription
+
+
+class AbstractFileSystem(Protocol):
+    """A file system interface."""
+
+    def localize_description(
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply bartender/filesystems/abstract.py:10]: Postponed https://github.com/i-VRESSE/bartender/issues/34

File: bartender/filesystems/sftp.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            # so /remote/jobid/output becomes /local/jobid/output,
+            # otherwise it would be /local/jobid/jobid/output
+            localpaths = [str(src.job_dir)]
+            remotepath = str(target.job_dir.parent)
+            await sftp.get(localpaths, remotepath, recurse=True)
+
+    def close(self) -> None:
+        """Close SSH connection."""
+        if self.conn:
+            self.conn.close()
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply bartender/filesystems/sftp.py:81]: Postponed to https://github.com/i-VRESSE/bartender/issues/35

**Pull request merged**